libs/auth/src/java_card.ts

```diff
@@ -0,0 +1,462 @@
+        CARD_VX_ADMIN_CERT.PRIVATE_KEY_SLOT,
+        cardVxAdminCert,
+        DEFAULT_PIN
+      );
+    }
+
+    return user;
+  }
+
+  private async selectApplet(): Promise<void> {
+    await this.cardReader.transmit(
+      new CommandApdu({
+        ins: SELECT.INS,
+        p1: SELECT.P1,
+        p2: SELECT.P2,
+        data: OPENFIPS201_AID,
+      })
+    );
+  }
+
+  /**
+   * Retrieves the specified cert in PEM format
+   */
+  private async retrieveCert(certObjectId: Buffer): Promise<Buffer> {
+    const getDataResult = await this.cardReader.transmit(
+      new CommandApdu({
+        ins: GET_DATA.INS,
+        p1: GET_DATA.P1,
+        p2: GET_DATA.P2,
+        data: constructTlv(GET_DATA.TAG_LIST_TAG, certObjectId),
+      })
+    );
+    const certInDerFormat = getDataResult.subarray(8, -5); // Trim metadata
+    const certInPemFormat = await openssl([
+      'x509',
+      '-inform',
+      'DER',
+      '-outform',
+      'PEM',
+      '-in',
+      certInDerFormat,
+    ]);
+    return certInPemFormat;
+  }
+
+  /**
+   * Verifies that the card private key in the specified slot corresponds to the public key in the
+   * provided cert by 1) having the private key sign a "challenge" and 2) using the public key to
+   * verify the generated signature.
+   *
+   * A PIN must be provided if operations with the specified private key are PIN-gated.
+   */
+  private async verifyCardPrivateKey(
+    privateKeySlot: Byte,
+    cert: Buffer,
+    pin?: string
+  ) {
+    if (pin) {
+      // Check the PIN
+      await this.cardReader.transmit(
+        new CommandApdu({
+          ins: VERIFY.INS,
+          p1: VERIFY.P1_VERIFY,
+          p2: VERIFY.P2_PIN,
+          data: construct8BytePinBuffer(pin),
+        })
+      );
+    }
+
+    // Have the private key in the specified slot sign a "challenge"
+    const challenge = new Date().toString();
+    const challengeHash = Buffer.from(sha256(challenge), 'hex');
+    const generalAuthenticateResponse = await this.cardReader.transmit(
+      new CommandApdu({
+        ins: GENERAL_AUTHENTICATE.INS,
+        p1: GENERAL_AUTHENTICATE.P1_ECC256,
+        p2: privateKeySlot,
+        data: constructTlv(
+          GENERAL_AUTHENTICATE.DYNAMIC_AUTHENTICATION_TEMPLATE_TAG,
+          Buffer.concat([
+            constructTlv(GENERAL_AUTHENTICATE.CHALLENGE_TAG, challengeHash),
+            constructTlv(GENERAL_AUTHENTICATE.RESPONSE_TAG, Buffer.from([])),
+          ])
+        ),
+      })
+    );
+    const challengeSignature = generalAuthenticateResponse.subarray(4); // Trim metadata
+
+    // Use the public key in the provided cert to verify the generated signature
+    const certPublicKey = await openssl([
+      'x509',
+      '-noout',
+      '-pubkey',
+      '-in',
+      cert,
+    ]);
+    await openssl([
+      'dgst',
+      '-verify',
+      certPublicKey,
+      '-sha256',
+      '-signature',
+      challengeSignature,
+      Buffer.from(challenge, 'utf-8'),
+    ]);
```

Comment:
Yes good q! This does error out if the verification fails. The exit status of the openssl shell command in this case is non-zero, which causes the convenience function's returned promise to reject:

https://github.com/votingworks/vxsuite/blob/fb695c7e7ffe5c1be2ea276f678563b32b62d022/libs/auth/src/openssl.ts#L61-L62

In case of success, the openssl shell command outputs "Verified OK" to the standard output, but I opted to check the exit status instead of the text output since that seemed more reliable. The convenience function's promise only resolves when the exit status is zero:

https://github.com/votingworks/vxsuite/blob/fb695c7e7ffe5c1be2ea276f678563b32b62d022/libs/auth/src/openssl.ts#L65-L66

Reply:
OK sounds good, can you add a comment here that explains, otherwise it feels like we're not really verifying and  it sets off the alarm bells.

Reply:
https://github.com/votingworks/vxsuite/pull/3037/commits/00b85194c131d875386c9d89afb81eebb44fc710 ✅